### PR TITLE
Remove a few `Rc`s from RegionInferenceCtxt

### LINKED ISSUE #55853: Possible enhancement for nll::region_infer::RegionInferenceContext
I noticed that `constraints`, `constraint_graph` and `universal_region_relations` (all members of the `RegionInferenceContext` struct) are wrapped in `Rc`s, but it doesn't seem necessary - none of them are `clone`d anywhere. I thought this might be done in order to reduce the size of the `RegionInferenceContext` struct, but the same could also be done to its other components that are bigger than `Rc` as well.

A comment in `region_infer::mod.rs` suggests that using an `Rc` "freezes constraints", but if they are just to be frozen a plain immutable reference could used instead, no?

cc @nnethercote

### PR DESCRIPTION
**Pull request by mark-i-m**
fixes https://github.com/rust-lang/rust/issues/55853

r? @matthewjasper

### CODE REVIEW

File: src/librustc_mir/borrow_check/mod.rs [hunk 1]
```diff
@@ -73,6 +73,24 @@ crate use place_ext::PlaceExt;
 crate use places_conflict::{places_conflict, PlaceConflictBias};
 crate use region_infer::RegionInferenceContext;
 
+/// An owned immutable value.
+#[derive(Debug)]
+struct Frozen<T>(T);
```
Comment (Centril) [inline src/librustc_mir/borrow_check/mod.rs]: Maybe this should be moved to rustc_data_structures?

File: src/librustc_data_structures/frozen.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
+//! An immutable, owned value.
```
Comment (Centril) [inline src/librustc_data_structures/frozen.rs]: Up to interior mutability (unless you want to use `Freeze` to exclude those).

Comment (mark-i-m) [reply src/librustc_data_structures/frozen.rs]: Yes, I updated the docs.

**Pull request merged**